### PR TITLE
[callouts] Add symbol-related variables to the expression context

### PR DESCRIPTION
**Pull request by nirvn**
## Description
@nyalldawson , this PR follows up your great callouts work to bring the @symbol_color variable to  callout (through the label feature object) rendering context.

This is how it looks when used against symbols from the rule-based renderer:
Screenshot from 2019-07-16 11-25-57

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: src/core/qgslabelingengine.cpp [hunk 1]
```diff
@@ -371,6 +373,10 @@ void QgsLabelingEngine::run( QgsRenderContext &context )
 
     context.expressionContext().setFeature( lf->feature() );
     context.expressionContext().setFields( lf->feature().fields() );
+    if ( lf->symbol() )
+    {
+      symbolScope = QgsExpressionContextUtils::updateSymbolScope( lf->symbol(), symbolScope );
+    }
     lf->provider()->drawLabel( context, label );
   }
```
Comment (nyalldawson) [inline src/core/qgslabelingengine.cpp]: Don't forget to pop the symbolScope here. You'll likely also need this in the new draw label background loop too.

File: src/core/qgspallabeling.h [hunk 2]
```diff
@@ -898,7 +899,7 @@ class CORE_EXPORT QgsPalLayerSettings
      */
     void registerFeature( const QgsFeature &f, QgsRenderContext &context,
                           QgsLabelFeature **labelFeature SIP_PYARGREMOVE = nullptr,
-                          QgsGeometry obstacleGeometry SIP_PYARGREMOVE = QgsGeometry() );
+                          QgsGeometry obstacleGeometry SIP_PYARGREMOVE = QgsGeometry(), const QgsSymbol *symbol SIP_PYARGREMOVE = nullptr );
```
Comment (nyalldawson) [inline src/core/qgspallabeling.h]: What happens to ownership here? Needs a note in the dox

File: src/core/qgsvectorlayerlabelprovider.cpp [hunk 3]
```diff
@@ -141,6 +141,7 @@ QList<QgsLabelFeature *> QgsVectorLayerLabelProvider::labelFeatures( QgsRenderCo
   while ( fit.nextFeature( fet ) )
   {
     QgsGeometry obstacleGeometry;
+    QgsSymbol *symbol;
```
Comment (nyalldawson) [inline src/core/qgsvectorlayerlabelprovider.cpp]: std::
```suggestion
    std::unique_ptr< QgsSymbol > symbol;
```

File: src/core/qgsvectorlayerlabelprovider.cpp [hunk 4]
```diff
@@ -151,11 +152,12 @@ QList<QgsLabelFeature *> QgsVectorLayerLabelProvider::labelFeatures( QgsRenderCo
       }
       if ( !symbols.isEmpty() )
       {
-        symbolScope = QgsExpressionContextUtils::updateSymbolScope( symbols.at( 0 ), symbolScope );
+        symbol = symbols.at( 0 )->clone();
```
Comment (nyalldawson) [inline src/core/qgsvectorlayerlabelprovider.cpp]:
```suggestion
        symbol.reset( symbols.at( 0 )->clone() );
```

File: src/core/qgsvectorlayerlabelprovider.cpp [hunk 5]
```diff
@@ -151,11 +152,12 @@ QList<QgsLabelFeature *> QgsVectorLayerLabelProvider::labelFeatures( QgsRenderCo
...
+        symbolScope = QgsExpressionContextUtils::updateSymbolScope( symbol, symbolScope );
       }
     }
     ctx.expressionContext().setFeature( fet );
-    registerFeature( fet, ctx, obstacleGeometry );
+    registerFeature( fet, ctx, obstacleGeometry, symbol );
```
Comment (nyalldawson) [inline src/core/qgsvectorlayerlabelprovider.cpp]:
```suggestion
    registerFeature( fet, ctx, obstacleGeometry, symbol.release() );
```

File: src/core/qgsvectorlayerlabelprovider.h [hunk 6]
```diff
@@ -75,8 +75,9 @@ class CORE_EXPORT QgsVectorLayerLabelProvider : public QgsAbstractLabelProvider
      * should be used as an obstacle for labels (e.g., if the feature has been rendered with an offset point
      * symbol, the obstacle geometry should represent the bounds of the offset symbol). If not set,
      * the feature's original geometry will be used as an obstacle for labels.
+     * \param symbol feature symbol to label
```
Comment (nyalldawson) [inline src/core/qgsvectorlayerlabelprovider.h]: Ownership of symbol is transferred

File: src/core/qgsvectorlayerlabelprovider.h [hunk 7]
```diff
@@ -75,8 +75,9 @@ class CORE_EXPORT QgsVectorLayerLabelProvider : public QgsAbstractLabelProvider
...
      */
-    virtual void registerFeature( const QgsFeature &feature, QgsRenderContext &context, const QgsGeometry &obstacleGeometry = QgsGeometry() );
+    virtual void registerFeature( const QgsFeature &feature, QgsRenderContext &context, const QgsGeometry &obstacleGeometry = QgsGeometry(), const QgsSymbol *symbol = nullptr );
```
Comment (nyalldawson) [inline src/core/qgsvectorlayerlabelprovider.h]:
```suggestion
    virtual void registerFeature( const QgsFeature &feature, QgsRenderContext &context, const QgsGeometry &obstacleGeometry = QgsGeometry(), const QgsSymbol *symbol SIP_TRANSFER = nullptr );
```

File: src/core/qgsvectorlayerrenderer.cpp [hunk 8]
```diff
@@ -309,20 +309,21 @@ void QgsVectorLayerRenderer::drawRenderer( QgsFeatureIterator &fit )
         {
           QgsGeometry obstacleGeometry;
           QgsSymbolList symbols = mRenderer->originalSymbolsForFeature( fet, mContext );
-
+          QgsSymbol *symbol;
```
Comment (nyalldawson) [inline src/core/qgsvectorlayerrenderer.cpp]: unique ptr

File: src/core/qgsvectorlayerrenderer.cpp [hunk 9]
```diff
@@ -392,6 +393,7 @@ void QgsVectorLayerRenderer::drawRendererLevels( QgsFeatureIterator &fit )
     if ( mContext.labelingEngine() )
     {
       QgsGeometry obstacleGeometry;
+      QgsSymbol *symbol;
```
Comment (nyalldawson) [inline src/core/qgsvectorlayerrenderer.cpp]: unique ptr

File: tests/src/core/testqgscallout.cpp [hunk 10]
```diff
@@ -534,6 +535,58 @@ void TestQgsCallout::calloutDataDefinedSymbol()
...
+
+  QgsLabelingEngine engine;
+  engine.setMapSettings( mapSettings );
+  engine.addProvider( new QgsVectorLayerLabelProvider( vl, QString(), true, &settings ) );
+  //engine.setFlags( QgsLabelingEngine::RenderOutlineLabels | QgsLabelingEngine::DrawLabelRectOnly );
+  engine.run( context );
+
+  p.end();
+
+  QVERIFY( imageCheck( "callout_data_defined_symbol_color", img, 20 ) );
```
Comment (nyalldawson) [inline tests/src/core/testqgscallout.cpp]: missing?

File: src/core/qgslabelfeature.h [hunk 11]
```diff
@@ -387,6 +387,22 @@ class CORE_EXPORT QgsLabelFeature
...
+     */
+    const QgsSymbol *symbol() { return mSymbol.get(); }
+
+    /**
+     * Sets the feature \a symbol associated with this label.
+     * \see symbol()
+     *
+     * \since QGIS 3.10
+     */
+    void setSymbol( const QgsSymbol *symbol ) { mSymbol.reset( symbol ); }
```
Comment (nyalldawson) [inline src/core/qgslabelfeature.h]:
```suggestion
    void setSymbol( const QgsSymbol *symbol SIP_TRANSFER ) { mSymbol.reset( symbol ); }
```

File: src/core/qgslabelfeature.h [hunk 12]
```diff
@@ -387,6 +387,22 @@ class CORE_EXPORT QgsLabelFeature
...
+     * Returns the feature symbol associated with this label.
+     * \see setSymbol()
+     *
+     * \since QGIS 3.10
+     */
+    const QgsSymbol *symbol() { return mSymbol.get(); }
+
+    /**
+     * Sets the feature \a symbol associated with this label.
+     * \see symbol()
```
Comment (nyalldawson) [inline src/core/qgslabelfeature.h]: Ownership of \a symbol is transferred to the label feature.

File: src/core/qgslabelingengine.cpp (hunk 1 above)
Comment (nirvn) [reply src/core/qgslabelingengine.cpp]: @nyalldawson , do you want the scope to be appended and popped before / after loop or at every iteration?

Comment (nyalldawson) [reply src/core/qgslabelingengine.cpp]: No, after the final iteration only

File: src/core/qgsrulebasedlabeling.cpp [hunk 13]
```diff
@@ -345,18 +345,22 @@ void QgsRuleBasedLabeling::Rule::prepare( QgsRenderContext &context, QSet<QStrin
   }
 }
 
-QgsRuleBasedLabeling::Rule::RegisterResult QgsRuleBasedLabeling::Rule::registerFeature( const QgsFeature &feature, QgsRenderContext &context, QgsRuleBasedLabeling::RuleToProviderMap &subProviders, const QgsGeometry &obstacleGeometry )
+QgsRuleBasedLabeling::Rule::RegisterResult QgsRuleBasedLabeling::Rule::registerFeature( const QgsFeature &feature, QgsRenderContext &context, QgsRuleBasedLabeling::RuleToProviderMap &subProviders, const QgsGeometry &obstacleGeometry, const QgsSymbol *symbol )
 {
+  Q_UNUSED( symbol )
```
Comment (nyalldawson) [inline src/core/qgsrulebasedlabeling.cpp]: Remove

File: src/core/qgsrulebasedlabeling.cpp [hunk 14]
```diff
@@ -345,18 +345,22 @@ void QgsRuleBasedLabeling::Rule::prepare( QgsRenderContext &context, QSet<QStrin
...
     return Filtered;
+  }
 
   bool registered = false;
 
   // do we have active subprovider for the rule?
   if ( subProviders.contains( this ) && mIsActive )
   {
-    subProviders[this]->registerFeature( feature, context, obstacleGeometry );
+    subProviders[this]->registerFeature( feature, context, obstacleGeometry, symbol->clone() );
```
Comment (nyalldawson) [inline src/core/qgsrulebasedlabeling.cpp]:
```suggestion
    subProviders[this]->registerFeature( feature, context, obstacleGeometry, symbol ? symbol->clone() : nullptr );
```

File: src/core/qgsrulebasedlabeling.cpp [hunk 15]
```diff
@@ -380,10 +384,12 @@ QgsRuleBasedLabeling::Rule::RegisterResult QgsRuleBasedLabeling::Rule::registerF
   {
     for ( Rule *rule : qgis::as_const( mElseRules ) )
     {
-      registered |= rule->registerFeature( feature, context, subProviders, obstacleGeometry ) != Filtered;
+      registered |= rule->registerFeature( feature, context, subProviders, obstacleGeometry, symbol->clone() ) != Filtered;
```
Comment (nyalldawson) [inline src/core/qgsrulebasedlabeling.cpp]:
```suggestion
      registered |= rule->registerFeature( feature, context, subProviders, obstacleGeometry, symbol ? symbol->clone() : nullptr ) != Filtered;
```

**Pull request merged**